src/compiler/checker.ts

```diff
@@ -26740,7 +26740,11 @@ export function createTypeChecker(host: TypeCheckerHost): TypeChecker {
     function hasMatchingArgument(expression: CallExpression | NewExpression, reference: Node) {
         if (expression.arguments) {
             for (const argument of expression.arguments) {
-                if (isOrContainsMatchingReference(reference, argument) || optionalChainContainsReference(argument, reference)) {
+                if (
+                    isOrContainsMatchingReference(reference, argument)
+                    || optionalChainContainsReference(argument, reference)
+                    || isAccessExpression(argument) && isMatchingReference(reference, argument.expression)

```

Comment:
Wouldn't it be better to `getDiscriminantPropertyAccess(argument, type)` so the argument only matches if it has an actual discriminable property? AFAIK, `hasMatchingArgument` is only used in two places, so specializing it a bit more for this usecase is probably fine. Technically `getDiscriminantPropertyAccess` can also match binding patterns and identifiers aliasing discriminable properties, too, eg, 

```ts
const kind = fruid.kind;
if (isOneOf(kind, ['apple', 'banana'] as const)) {
    fruit.kind
    fruit
}
```

Reply:
We'll end up calling `getDiscriminantPropertyAccess` twice, but I'll try doing that.

Reply:
You know what, at this rate, if we need to call `getDiscriminantPropertyAccess` to decide whether we should call `narrowTypeByTypePredicate`, which will then possibly call `getDiscriminantPropertyAccess` (and also call `isMatchingReference` anyways), then we might as well just always call `narrowTypeByTypePredicate`? I'll try that and see if it affects performance.

Reply:
@weswigham I've switched to using `getCandidateDiscriminantPropertyAccess` for the above reasons.
Can you take a look again?